zeus/vcs/api.py

```diff
@@ -0,0 +1,197 @@
+
+    bits = auth_header.split(" ")
+    if len(bits) != 2:
+        return None
+
+    if bits[0].lower() != "token":
+        return None
+
+    return bits[1]
+
+
+def api_request(func):
+    @wraps(func)
+    def wrapper(request, *args, **kwargs):
+        @log_errors
+        async def tmp():
+            repo_id = request.query.get("repo_id")
+            if not repo_id:
+                return json_response({"error": "missing_arg"}, status=403)
+
+            token = get_token(request)
+            if not token:
+                return json_response({"error": "missing_auth"}, status=401)
+
+            token = auth.parse_token(token)
+            if not token:
+                current_app.logger.debug(
+                    "vcs-server.invalid-request command=%s reason=invalid-token",
+                    func.__name__,
+                )
+                return json_response({"error": "invalid_auth"}, status=401)
+
+            if "uid" in token:
+                with sentry_sdk.configure_scope() as scope:
+                    scope.user = {"id": token["uid"]}
+
+            if "repo_ids" not in token or repo_id not in token["repo_ids"]:
+                current_app.logger.debug(
+                    "vcs-server.invalid-request command=%s tenant=%s reason=invalid-repo",
+                    func.__name__,
+                    token,
+                )
+                return json_response({"error": "access_denied"}, status=400)
+
+            current_app.logger.debug(
+                "vcs-server.request repo_id=%s command=%s tenant=%s",
+                repo_id,
+                func.__name__,
+                token,
+            )
+
+            try:
+                return await func(request, repo_id, *args, **kwargs)
+            except InvalidPublicKey:
+                current_app.logger.exception(
+                    "vcs-server.invalid-pubkey repo_id=%s", repo_id
+                )
+                return json_response({"error": "invalid_pubkey"}, status=400)
+            except Exception:
+                current_app.logger.exception(
+                    "vcs-server.unhandled-error repo_id=%s", repo_id
+                )
+                return json_response({"error": "unknown_error"}, status=500)
+
+        return tmp()
+
+    return wrapper
+
+
+@span("health_check")
+@log_errors
+async def health_check(request):
+    return Response(text='{"ok": true}')
+
+
+@span("stmt.log")
+@api_request
+@log_errors
+async def stmt_log(request, repo_id):
+    vcs = await get_vcs(request.app, repo_id)
+
+    queue = request.app["queue"]
+
+    parent = request.query.get("parent")
+    branch = request.query.get("branch")
+    offset = int(request.query.get("offset") or 0)
+    limit = int(request.query.get("limit") or 100)
+
+    try:
+        log_results = vcs.log(parent=parent, branch=branch, offset=offset, limit=limit)
+    except UnknownRevision:
+        # we're running a lazy update here if it didnt already exist
```

Comment:
this would suggest new commits, so maybe we should only attempt to save commits in this scenario?